cmd/run.go

```diff
@@ -83,146 +125,193 @@ Default builder is pack.
-	// mocks when testing for builder and runner)
-	client, done := newClient(ClientConfig{Verbose: cfg.Verbose}, fn.WithRegistry(cfg.Registry), fn.WithBuilder(builder))
+	client, done := newClient(ClientConfig{Verbose: cfg.Verbose}, o...)
 	defer done()
 
-	// Build?
-	// If --build was set to 'auto', only build if client detects the function
-	// is stale (has either never been built or has had filesystem modifications
-	// since the last build).
-	if cfg.Build == "auto" {
-		if !function.Built() {
-			if function, err = client.Build(cmd.Context(), function); err != nil {
-				return
-			}
-			if err = function.Write(); err != nil {
-				return err
-			}
-		}
-		fmt.Println("Function already built.  Use --build to force a rebuild.")
-		// Otherwise, --build should parse to a truthy value which indicates an explicit
-		// override.
-	} else {
-		build, err := strconv.ParseBool(cfg.Build)
-		if err != nil {
-			return fmt.Errorf("unrecognized value for --build '%v'.  accepts 'auto', 'true' or 'false' (or similarly truthy value)", build)
-		}
-		if build {
-			if function, err = client.Build(cmd.Context(), function); err != nil {
-				return err
-			}
-			if err = function.Write(); err != nil {
-				return err
-			}
-		} else {
-			fmt.Println("Function build disabled.")
+	// Build
+	//
+	// If requesting to run via the container, build the container if it is
+	// either out-of-date or a build was explicitly requested.
+	if cfg.Container && shouldBuild(cfg.Build, f, client) {
+		if f, err = client.Build(cmd.Context(), f); err != nil {
+			return
 		}
-
 	}
 
-	// Run the function at path
-	job, err := client.Run(cmd.Context(), function)
+	// Run
+	//
+	// Runs the code either via a container or the default host-based runniner.
+	// For the former, build is required and a container runtime.  For the
+	// latter, scaffolding is first applied and the local host must be
+	// configured to build/run the language of the function.
+	job, err := client.Run(cmd.Context(), f)
 	if err != nil {
 		return
 	}
 	defer job.Stop()
 
-	fmt.Fprintf(cmd.OutOrStderr(), "Function started on port %v\n", job.Port)
+	fmt.Fprintf(cmd.OutOrStderr(), "Running on host port %v\n", job.Port)
 
 	select {
 	case <-cmd.Context().Done():
 		if !errors.Is(cmd.Context().Err(), context.Canceled) {
 			err = cmd.Context().Err()
 		}
-		return
 	case err = <-job.Errors:
-		return
 	}
+
+	// NOTE: we do not f.Write() here unlike deploy (and build).
+	// running is ephemeral: a run is not affecting the function itself,
+	// as opposed to deploy commands, which are actually mutating the current
+	// state of the function as it exists on the network.
+	// Another way to think of this is that runs are development-centric tests,
+	// and thus most likely values changed such as environment variables,
+	// builder, etc. would not be expected to persist and affect the next deploy.
+	// Run is ephemeral, deploy is persistent.
```

Comment:
I wonder if this should somehow be communicated to the user.

Reply:
Yes and at least one time when I was using `run` locally, specifying `--builder`, I was confused why it didn't use the same value the next time around.

So on the one hand I am pretty sure that we do not want invocations of `run` to persist the value of flags which alter `deploy` (and `build`), but I also find that once accustomed to the value of flags persisting between commands, I expect it to be consistently available.

The only solution that comes to mind is that we actually have a separate set of flag values for running (a development/testing, inner-loop sort of paradigm) vs those used for the deployed function. These would be developer-specific (host scoped) and reside in `.func/` in the same way as `--remote`, etc.

This is definitely complex enough to warrant its own issue and PR, so for the purposes of this PR I just want to make sure that invoking `run` with a different builder during development/testing doesn't alter the deployed function configuration!